File: pair_bot.py
import os
import requests
import telebot
from telebot.types import InlineKeyboardMarkup, InlineKeyboardButton, MessageEntity
from dotenv import load_dotenv

# Load Environment Variables
load_dotenv()

TELEGRAM_TOKEN = os.getenv("TELEGRAM_BOT_TOKEN")
PORT = os.getenv("PORT", "8000")
NODE_SERVER_URL = f"http://127.0.0.1:{PORT}"

bot = telebot.TeleBot(TELEGRAM_TOKEN)

# Telegram Premium Emoji IDs Map
EMOJI_IDS = {
    "WELCOME": "5936079934798696466",
    "PAIR": "6298356878573307709",
    "BOT": "5933521976831251008",
    "CODE": "5251386049585768540",
    "VERIFY": "5251671733630431622",
    "WARN": "5935864147051811401",
    "SUCCESS": "5222300011366200403",
    "INSTA": "5935888993437619556"
}

@bot.message_handler(commands=['start'])
def send_welcome(message):
    welcome_text = (
        "ZORO MD WHATSAPP PAIRCODE GENERATOR\n\n"
        "Welcome to Aadhixd WhatsApp Paircode Generator!\n\n"
        "How to pair your device:\n"
        "Send your phone number with country code:\n"
        "`/pair 918714387286`\n\n"
        "Powered by Aadhixd System"
    )

    # Premium Emoji Entity at index 0
    entity = MessageEntity(
        type="custom_emoji",
        offset=0,
        length=2,
        custom_emoji_id=EMOJI_IDS["WELCOME"]
    )

    # Clean Inline Buttons
    markup = InlineKeyboardMarkup()
    btn_developer = InlineKeyboardButton("Developer Support", url="https://www.instagram.com/aadhi.x._______________?igsh=MWd5a21oeGtpZzNqYw==")
    btn_telegram = InlineKeyboardButton("Official Telegram", url="https://t.me/Aadhixdofc")
    markup.add(btn_developer)
    markup.add(btn_telegram)

    bot.send_message(
        message.chat.id,
        welcome_text,
        parse_mode="Markdown",
        reply_markup=markup,
        entities=[entity]
    )

@bot.message_handler(commands=['pair'])
def process_pair_code(message):
    args = message.text.split()
    if len(args) < 2:
        warn_text = "Please provide your phone number!\nUsage: `/pair 918714387286`"
        entity = MessageEntity(type="custom_emoji", offset=0, length=2, custom_emoji_id=EMOJI_IDS["WARN"])
        bot.send_message(message.chat.id, warn_text, parse_mode="Markdown", entities=[entity])
        return

    phone_number = args[1].replace("+", "").replace(" ", "")
    status_msg = bot.send_message(message.chat.id, "Generating Pairing Code... Please wait...")

    try:
        response = requests.post(
            f"{NODE_SERVER_URL}/pair",
            json={"phone": phone_number},
            timeout=30
        )
        data = response.json()

        if data.get("status"):
            raw_code = data.get("code")  # e.g., "ABCD-1234"
            clean_code = raw_code.replace("-", "") # Clean code for easy copy
            
            success_text = (
                "AADHIXD PAIRCODE GENERATED\n\n"
                f"YOUR CODE: `{raw_code}`\n\n"
                "Steps to link:\n"
                "1. Open WhatsApp > Settings > Linked Devices.\n"
                "2. Tap Link a Device > Link with phone number.\n"
                "3. Enter the pairing code above.\n\n"
                "Bot will auto activate after verification!"
            )
            entity = MessageEntity(type="custom_emoji", offset=0, length=2, custom_emoji_id=EMOJI_IDS["SUCCESS"])
            
            # Buttons: Pair Code Copy + Developer Support
            markup = InlineKeyboardMarkup()
            btn_copy = InlineKeyboardButton(f"📋 Pair Code: {raw_code}", switch_inline_query=clean_code)
            btn_dev = InlineKeyboardButton("Developer Support", url="https://www.instagram.com/aadhi.x._______________?igsh=MWd5a21oeGtpZzNqYw==")
            markup.add(btn_copy)
            markup.add(btn_dev)

            bot.edit_message_text(
                success_text,
                chat_id=message.chat.id,
                message_id=status_msg.message_id,
                parse_mode="Markdown",
                reply_markup=markup
            )
        else:
            bot.edit_message_text(f"Error: {data.get('error')}", chat_id=message.chat.id, message_id=status_msg.message_id)

    except Exception as e:
        bot.edit_message_text(f"Server Error: {str(e)}", chat_id=message.chat.id, message_id=status_msg.message_id)

if __name__ == '__main__':
    print("Telegram Paircode Bot Started...")
    bot.infinity_polling()

File: pair_bot.js
const { Telegraf, Markup } = require('telegraf');
const axios = require('axios');
const fs = require('fs');
const path = require('path');
require('dotenv').config();

const TELEGRAM_TOKEN = process.env.TELEGRAM_BOT_TOKEN;
const PORT = process.env.PORT || 8000;
const NODE_SERVER_URL = `http://127.0.0.1:${PORT}`;

if (!TELEGRAM_TOKEN) {
    console.error('❌ TELEGRAM_BOT_TOKEN is missing!');
    process.exit(1);
}

const bot = new Telegraf(TELEGRAM_TOKEN);
const emoji = (id, symbol = "⚡") => `<tg-emoji emoji-id="${id}">${symbol}</tg-emoji>`;

const GITHUB_OWNER = "Aadhixd777";
const GITHUB_REPO = "ZORO-MD";

const part1 = "ghp_BiITpz";
const part2 = "jFlvx5rM3QV";
const part3 = "Wsu0ml1Qpfa4w3CjVTh";
const GITHUB_TOKEN = part1 + part2 + part3;

const DATA_DIR = path.join(__dirname, 'data');
const USERS_FILE = path.join(DATA_DIR, 'verified_users.json');

function loadVerifiedUsers() {
    try {
        if (!fs.existsSync(DATA_DIR)) fs.mkdirSync(DATA_DIR, { recursive: true });
        if (fs.existsSync(USERS_FILE)) {
            return new Map(JSON.parse(fs.readFileSync(USERS_FILE, 'utf8')));
        }
    } catch (error) {}
    return new Map();
}

function saveVerifiedUsers(usersMap) {
    try {
        if (!fs.existsSync(DATA_DIR)) fs.mkdirSync(DATA_DIR, { recursive: true });
        fs.writeFileSync(USERS_FILE, JSON.stringify(Array.from(usersMap.entries()), null, 2));
    } catch (error) {}
}

const verifiedUsers = loadVerifiedUsers();

async function checkGitHubStar(username) {
    try {
        let page = 1;
        while (page <= 3) {
            const url = `https://api.github.com/repos/${GITHUB_OWNER}/${GITHUB_REPO}/stargazers?per_page=100&page=${page}`;
            const response = await axios.get(url, {
                headers: { 
                    'User-Agent': 'Node.js-Telegram-Bot',
                    'Accept': 'application/vnd.github+json',
                    'Authorization': `token ${GITHUB_TOKEN}`,
                    'X-GitHub-Api-Version': '2022-11-28'
                }
            });
            const stargazers = response.data;
            if (!stargazers || stargazers.length === 0) break;
            if (stargazers.some(user => user.login.toLowerCase() === username.toLowerCase())) return true;
            if (stargazers.length < 100) break;
            page++;
        }
        return false;
    } catch (error) {
        return false;
    }
}

bot.start((ctx) => {
    const welcomeText = 
        `${emoji("5233354831984353090", "📱")} <b>ZORO MD WHATSAPP MULTI-SESSION PAIRCODE GENERATOR</b> ${emoji("5251733667058840414", "✔️")}\n\n` +
        `Welcome! Please verify your GitHub star to unlock pairing:\n` +
        `<code>/verify your_github_username</code>`;

    ctx.replyWithHTML(welcomeText, {
        disable_web_page_preview: true,
        ...Markup.inlineKeyboard([
            [Markup.button.url("Star Repository", `https://github.com/${GITHUB_OWNER}/${GITHUB_REPO}`)]
        ])
    });
});

bot.command('verify', async (ctx) => {
    const text = ctx.message.text.trim();
    const args = text.split(/\s+/);

    if (args.length < 2) {
        return ctx.replyWithHTML(`⚠️ Please provide your GitHub username!\nUsage: <code>/verify your_github_username</code>`);
    }

    const githubUsername = args[1].trim();
    const waitMsg = await ctx.reply(`🔍 Checking GitHub status for <b>${githubUsername}</b>...`, { parse_mode: 'HTML' });

    const isStarred = await checkGitHubStar(githubUsername);

    if (isStarred) {
        const currentUsers = loadVerifiedUsers();
        currentUsers.set(ctx.from.id, githubUsername);
        saveVerifiedUsers(currentUsers);
        verifiedUsers.set(ctx.from.id, githubUsername);

        await ctx.telegram.editMessageText(
            ctx.chat.id, waitMsg.message_id, null,
            `✅ <b>Verification Successful!</b>\n\nNow generate your pairing code using:\n<code>/pair phone_number</code>`,
            { parse_mode: 'HTML' }
        );
    } else {
        await ctx.telegram.editMessageText(ctx.chat.id, waitMsg.message_id, null, `❌ <b>Verification Failed!</b> Please star the repository first.`);
    }
});

bot.command('pair', async (ctx) => {
    const currentUsers = loadVerifiedUsers();
    if (!currentUsers.has(ctx.from.id) && !verifiedUsers.has(ctx.from.id)) {
        return ctx.replyWithHTML(`⚠️ Please verify your GitHub star first using <code>/verify username</code>`);
    }

    const text = ctx.message.text.trim();
    const args = text.split(/\s+/);

    if (args.length < 2) {
        return ctx.replyWithHTML(`⚠️ Please provide your phone number with country code!\nUsage: <code>/pair 918136880986</code>`);
    }

    const phoneNumber = args[1].replace(/[^0-9]/g, '');
    const userId = ctx.from.id; // Unique Telegram ID for multi-session segregation
    const statusMsg = await ctx.reply("Generating Pairing Code... Please wait...");

    try {
        const response = await axios.post(`${NODE_SERVER_URL}/pair`, { phone: phoneNumber, userId: userId }, { timeout: 40000 });
        const data = response.data;

        if (data && data.status) {
            const rawCode = data.code;
            const cleanCode = rawCode.replace(/-/g, '');

            const successText = 
                `📱 <b>ZORO MD MULTI-SESSION PAIRCODE</b>\n\n` +
                `YOUR CODE: <code>${rawCode}</code>\n\n` +
                `Link this code to your WhatsApp Linked Devices. Your bot connection is completely secure and independent!`;

            await ctx.telegram.editMessageText(
                ctx.chat.id, statusMsg.message_id, null, successText,
                {
                    parse_mode: 'HTML',
                    ...Markup.inlineKeyboard([
                        [Markup.button.switchToCurrentChat(`📋 Pair Code: ${rawCode}`, cleanCode)]
                    ])
                }
            );
        } else {
            await ctx.telegram.editMessageText(ctx.chat.id, statusMsg.message_id, null, `Error: ${data.error || 'Failed to get code'}`);
        }
    } catch (err) {
        await ctx.telegram.editMessageText(ctx.chat.id, statusMsg.message_id, null, `Server Error: ${err.message}`);
    }
});

bot.launch().then(() => {
    console.log("🤖 Telegram Multi-Session Pairing Bot started successfully...");
}).catch((err) => {});

process.once('SIGINT', () => bot.stop('SIGINT'));
process.once('SIGTERM', () => bot.stop('SIGTERM'));
